cosi/cosi.md

```diff
@@ -258,64 +258,82 @@ steps below successfully.
 
 # Collective Signing Protocol 
 
-This section introduces the distributed CoSi protocol with n participants. For
-simplicity, we assume there is a designated leader who is
-responsible for collecting the shares and generating the signature. This leader
-could be any of the signers and is not trusted in any way. All participants are
-communicating through a reliable channel with the leader.
+This section introduces the distributed CoSi protocol with n participants. 
+The CoSi protocol has four distinct phases, and each one of these phases uses a
+distinct packet:
++ the Announcement packet for the Announcement phase
++ the Commitment packet for the Commitment phase 
++ the Challenge packet for the Challenge phase 
++ the Response packet for the Response phase 
+
+The packet formats are described in details in the section XXX Packet Format XXX.
+
+For simplicity, we assume there is a designated leader who is responsible for
+collecting the shares and generating the signature. This leader could be any of
+the signers and is not trusted in any way. All participants are communicating
+through a reliable channel with the leader.
 
 ## Collective Signature
 
-The leader must know the statement S to be signed and the set of public keys of
-the participants N. The point A is defined as the collective key of the
-participants N. A collective signature is generated in four steps over two
+The leader must know the message M to be signed and the set of public keys of
+the participants P. The point A is defined as the collective key of the
+participants P. A collective signature is generated in four steps over two
 round trips between the leader and the rest of the participants. 
 
 ### Announcement
 
-Upon the request to generate a signature on a statement S, the leader broadcasts
-an announcement message indicating the start of a signing process. It is up to
-the implementation to decide whether to send S itself during that phase or not. 
+Upon the request to generate a signature on a message M, the leader broadcasts
+an Announcement packet indicating the start of a signing process. This
+Announcement packet MUST contain the message M to sign.
```

Comment:
No, I think this is overly strong and not strictly necessary, and likely undesirable since it's likely to eliminate important pipelining opportunities.

I understand your concern that without knowing the message they're supposed to sign beforehand, cosigners that don't like the message will only be able to decide this in the response phase, and hence will require the protocol to restart (in the absence of an exception mechanism, which we haven't defined yet).  However, IF this proves to be a problem, then a better solution might be to add an exception mechanism rather than to add a strict requirement that the message be known in the announcement phase.

Also, I'm not convinced it's really likely to be a problem: in many/most scenarios I think we care about, the cosigners are typically supposed to be verifying a deterministic set of requirements that "should" be exactly the same as what the leader and every other cosigner are verifying themselves.  This is the case for blockchains (where everyone is supposed to agree on the validity rules), etc.  If an honest cosigner gets asked by a leader to sign something that doesn't check out, then that's probably an indication that the leader is seriously buggy or malicious, and it's perfectly reasonable for that cosigner to bail in a not-particularly-graceful way.  (And if the cosigner is dishonest and wants to DoS attack the signing process, we know the dishonest cosigner can do that regardless unless we jump into the added complexity of a restart-free scheme.)

In other words, for most of the applications I think we care about, cosigning should be a fully-automated process - it's not about some human sitting there navel-gazing and figuring out if they're comfortable cosigning or not, but about the cosigner simply double-checking that the thing the leader is asking to sign is indeed consistent with all the rules the leader itself should already have imposed on the message.  Being able to "gracefully decline" cosigning is not necessarily a priority, until/unless we actually identify some application for which that mode actually matters.